File: wasm/src/TecQuery.hs
module TecQuery where

import Data.Aeson
  ( FromJSON,
    ToJSON (toEncoding),
    defaultOptions,
    genericToEncoding,
  )
import GHC.Generics (Generic)
import Optics.Core

data TecIndexSeq
  = TecIndexValues {tecIndexValues :: [String]}
  | TecIndexRange {tecIndexFrom :: String, tecIndexTo :: Maybe String}
  deriving (Show, Generic)

instance ToJSON TecIndexSeq where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecIndexSeq

data TecQuery = TecQuery
  { tecQueryFunc :: String,
    tecQueryIndexSeqs :: [TecIndexSeq]
  }
  deriving (Show, Generic)

instance ToJSON TecQuery where
  toEncoding = genericToEncoding defaultOptions

instance FromJSON TecQuery

_tecQueryIndexSeqs :: Lens' TecQuery [TecIndexSeq]
_tecQueryIndexSeqs = lens tecQueryIndexSeqs (\(TecQuery {tecQueryFunc}) tecQueryIndexSeqs -> TecQuery {tecQueryFunc, tecQueryIndexSeqs})

File: web/libs/codec/src/TecQuery.ts
import { Schema as S } from "effect";
import * as E from "effect";
export const TecIndexValues = S.Struct({
  tag: S.tag("TecIndexValues"),
  tecIndexValues: S.NonEmptyArray(S.String),
});
export type TecIndexValues = typeof TecIndexValues.Type;
export const TecIndexRange = S.Struct({
  tag: S.tag("TecIndexRange"),
  tecIndexFrom: S.String,
  tecIndexTo: S.Null,
});
export type TecIndexRange = typeof TecIndexRange.Type;
export const TecIndexSeq = S.Union(TecIndexValues, TecIndexRange);
export type TecIndexSeq = typeof TecIndexSeq.Type;
export const TecQuery = S.Struct({
  tecQueryFunc: S.String,
  tecQueryIndexSeqs: S.NonEmptyArray(TecIndexSeq),
});
export type TecQuery = typeof TecQuery.Type;
